File: exercises/design_dictionary_211.java
// Optimal Recursive Solution
class WordDictionary {
    private WordDictionary[] children;
    boolean isEndOfWord;
    public WordDictionary() {
        children = new WordDictionary[26];
    }
    public void addWord(String word) {
        WordDictionary curr = this;
        for(int i = 0; i < word.length(); i++){
            int c = word.charAt(i) - 'a';
            if(curr.children[c] == null)
                curr.children[c] = new WordDictionary();
            curr = curr.children[c];
        }
        curr.isEndOfWord = true;
    }
    public boolean search(String word) {
        WordDictionary curr = this;
        for(int i = 0; i < word.length(); ++i){
            char c = word.charAt(i);
            if(c == '.'){
                for(WordDictionary ch: curr.children)
                    if(ch != null && ch.search(word.substring(i+1))) return true;
                return false;
            }
            if(curr.children[c - 'a'] == null) return false;
            curr = curr.children[c - 'a'];
        }
        return curr != null && curr.isEndOfWord;
    }
}
// Original Ugly Iterative Solution
//class WordDictionary {
//    WordDictionary[] children;
//	boolean isEnd;
//	WordDictionary wildcard;
//    public WordDictionary() {
//		children = new WordDictionary[26];
//    }
//
//    public void addWord(String word) { _addWord(word, 0); }
//    private void _addWord(String word, int ind) {
//        if (ind >= word.length()) return;
//        int i = word.charAt(ind) - 'a';
//        if (this.children[i] == null) this.children[i] = new WordDictionary();
//        if (ind == word.length() - 1) this.children[i].isEnd = true;
//        this.children[i]._addWord(word, ind + 1);
//    }
//
//        public boolean search(String word) { 
//		Queue<WordDictionary> searches = new LinkedList<WordDictionary>();
//		searches.offer(this);
//		return _search(word, searches); 
//	}
//	private boolean _search(String word, Queue<WordDictionary> searches) {
//		boolean found = false;
//        boolean isEnd = false;
//		for (int i = 0; i < word.length(); i++) {
//            isEnd = false;
//            found = false;
//			if (word.charAt(i) == '.') {
//                int queueSize = searches.size();
//				for (int j = 0; j < queueSize; j++) {
//					WordDictionary currWordDictionary = searches.poll();
//					for (int k = 0; k < 26; k++) {
//						if (currWordDictionary.children[k] != null) {
//							searches.offer(currWordDictionary.children[k]);
//                            if (currWordDictionary.children[k].isEnd) isEnd = true;
//							found = true;
//						}
//					}
//				}
//			} else {
//                int queueSize = searches.size();
//				for (int j = 0; j < queueSize; j++) {
//					WordDictionary currWordDictionary = searches.poll();
//					int ch = word.charAt(i) - 'a';
//					if (currWordDictionary.children[ch] != null) {
//						searches.offer(currWordDictionary.children[ch]);
//                        if (currWordDictionary.children[ch].isEnd) isEnd = true;
//						found = true;
//					}
//				}
//			}
//		}
//        if (!found || !isEnd) return false;
//		return searches.size() > 0;
//	}
//
//}

File: exercises/design_dictionary_211.ts
// Optimal recursive Solution
class WordDictionary {
	children: WordDictionary[]
	isEnd: boolean = false
    constructor() {
		this.children = new Array(26);
    }

    addWord(word: string): void {
		let curr: WordDictionary = this
		for (let i = 0; i < word.length; i++) {
			const ind = word.charCodeAt(i) - 'a'.charCodeAt(0);
			if (curr.children[ind] == null) curr.children[ind] = new WordDictionary();
			curr = curr.children[ind]
		}
		curr.isEnd = true
    }

    search(word: string): boolean {
		return this._search(word, 0);
    }
	private _search(word: string, start: number): boolean {
		let curr: WordDictionary = this!
		for (let i = start; i < word.length; i++) {
			const ind = word.charCodeAt(i) - 'a'.charCodeAt(0)
			if (word[i] === ".") {
				for (const c of curr.children) {
					if (c != null && c._search(word, i+1)) return true
				}
                return false
			}
			if (curr.children[ind] == null) return false
			curr = curr.children[ind]
		}
		return curr != null && curr.isEnd
	}
}

